### PR TITLE
Barcode logic updates

### PR DESCRIPTION
**Pull request by nettej17**
Acceptance Criteria:
Accurately parse the following GS1 Application Identifiers from GS1 code-128 and data matrix barcodes: GTIN (01), Lot (10), Expiration date (17), Serial Number (23) and Item Count (30) 
Accurately parse regardless of which order the AIs appear in.
Support the <GS> control character which signifies the end of variable length applications identifiers when they are not the max length or the last AI in the string
Alphanumeric support needed for Lot and Serial application identifiers (the other AI’s on the list and UPC-A’s technically should only be numbers).
Remove parentheses if the manufacturer encoded them even though they are not GS1 compliant.
If the 14 character GTIN starts with 003, use characters 4-13 as the NDC
If a 12 digit UPC-A barcode and the UPC number starts with 3, use characters 2-11 as the NDC
Otherwise just return the full string scanned if not identified using one of the other regular expression matches already in the code (e.g. poptags)

### CODE REVIEW

File: lib/bark.js [hunk 1]
```diff
@@ -10,18 +10,18 @@
...
-    if (typeof input == 'string') {
-        this.setConfig(optConfig);
+    if(typeof input == 'string')
+    {
         this.parse(input);
-    } else {
-        this.setConfig(optConfig);
+        if(typeof optConfig == 'object') this.setConfig(optConfig);
     }
+    else if(typeof input == 'object') this.setConfig(input);
```
Comment (cameronf) [inline lib/bark.js]: this feels like it is in conflict with the last PR - now I'm confused as to which one modifies which...

File: lib/bark.js [hunk 2]
```diff
@@ -117,78 +159,100 @@ Bark.prototype.AIparsers = {
...
-Bark.prototype.parse = function(inputVal) {
+Bark.prototype.parse = function( inputVal ) {
 
     // Resetting things.
     this.input = inputVal;
+    inputVal = inputVal.replace(/[()]/g, '');
     this.init();
 
-    var hasLeadingZero = inputVal.substring(0, 1) == '0';
+    console.log('Parsing input.', this.input);
```
Comment (cameronf) [inline lib/bark.js]: Is this intentional

Comment (cameronf) [reply lib/bark.js]: along with the other console.logs below

Comment (nettej17) [reply lib/bark.js]: they exist in the original bower component, but I think Eric removed them, so I should probably go with his changes and remove them

File: lib/bark.js [hunk 3]
```diff
@@ -53,42 +54,69 @@ Bark.prototype.setConfig = function(overrides) {
     this.config = computed;
 }
 
+//Constructor to handle variable length AIs
+function VariableLengthAI(inputVal, startPos, aiMaxLength){
+    var gsChar = '<';
```
[the remark this answers is not in the document]
Comment (ramseychambers) [reply lib/bark.js]: Jeanette was still figuring out how the barcode scanner outputs the group separator. gsChar and gsLength were created as variables so that we could easily update it once we figure it out.

File: lib/bark.js [hunk 4]
```diff
@@ -53,42 +54,69 @@ Bark.prototype.setConfig = function(overrides) {
...
         this.AIs['SSCC'] = inputVal.substr(startPos + 2, 18);
-        return startPos + 2 + 18;
+        if(this.AIs)
+            return startPos + 2 + 18;
     },
-    '01': function(inputVal, startPos) {
+    '01': function(inputVal, startPos){
         this.AIs['GTIN'] = inputVal.substr(startPos + 2, 14);
-        this.AIs['NDC'] = inputVal.substr(startPos + 5, 10);
+        if (this.AIs['GTIN'].substr(0,3) === '003' || this.AIs['GTIN'].substr(0,3) === '103'  ){
```
[the remark this answers is not in the document]
Comment (ramseychambers) [reply lib/bark.js]: lines 87 and 88 shouldn't have been ported over. Leave Brody's code as is.

File: lib/bark.js [hunk 5]
```diff
@@ -53,42 +54,69 @@ Bark.prototype.setConfig = function(overrides) {
...
+            this.AIs['NDC'] = this.AIs['GTIN'].substr(3,10);
```
[the remark this answers is not in the document]
Comment (ramseychambers) [reply lib/bark.js]: This shouldn't have been ported over.

File: lib/bark.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (nettej17) [reply lib/bark.js]: Nope, it should not, there are other options than those two

File: lib/bark.js [hunk 6]
```diff
@@ -117,78 +159,100 @@ Bark.prototype.AIparsers = {
...
+        else if(inputVal.length == 13)
+        {
             this.type = 'EAN-13';
-        } else if (inputVal.length == 14) {
+        }
+        else if(inputVal.length == 14)
+        {
             this.type = 'ITF-14';
         }
-    } else {
```
Comment (ramseychambers) [inline lib/bark.js]: I didn't intentionally leave this out. If Brody included this for a reason it should stay.

File: lib/bark.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (nettej17) [reply lib/bark.js]: Yes, this needs to be addressed

File: lib/bark.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (nettej17) [reply lib/bark.js]: I don't know, the original Bark-js code uses this magic number 2 (for the AI string of two characters) all over the place in the weirdest logical way where it's hard to follow, but in the end you somehow end up with the results you want. So this works in testing...

File: lib/bark.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (nettej17) [reply lib/bark.js]: wait so we decided to drop the 103 thing?

File: lib/bark.js [hunk 7]
```diff
@@ -117,78 +159,100 @@ Bark.prototype.AIparsers = {
...
+            else
+            {
                 // The AI parser will return the end position of its data
                 startPos = this.AIparsers[guessAI].call(this, inputVal, startPos);
                 currPos = 1;
             }
         }
 
         // Check if we parsed everything there was to parse
-        if (startPos != inputVal.length) {
```
Comment (ramseychambers) [inline lib/bark.js]: Didn't intentionally remove this. Not sure how startPos < inputVal.length vs startPos != inputVal.length impacts how it works.

File: bower.json [hunk 8]
```diff
@@ -1,11 +1,11 @@
 {
   "name": "bark-js",
   "version": "0.0.2",
-  "homepage": "https://github.com/Sleavely/Bark-js",
+  "homepage": "https://github.com/kitcheck/Bark-JS",
```
Comment (ramseychambers) [inline bower.json]: I didn't intend for this to change. May Mike did.

File: bower.json [hunk 9]
```diff
@@ -1,11 +1,11 @@
...
   "description": "Parse barcode inputs into a unified GS1-128 like format",
   "repository": {
     "type": "git",
-    "url": "https://github.com/Sleavely/Bark-JS"
+    "url": "https://github.com/kitcheck/Bark-JS"
```
Comment (ramseychambers) [inline bower.json]: I didn't intend for this to change. Maybe Mike did.

File: lib/bark.js [hunk 10]
```diff
@@ -117,78 +159,100 @@ Bark.prototype.AIparsers = {
...
+        if(inputVal.length == 11)
+        {
             this.type = 'UPC-A';
             //TODO: The check digit may have been omitted. should we attempt to add it ourselves?
-        } else if (inputVal.length == 12) {
+        }
+        else if(inputVal.length == 12)
+        {
             this.type = 'UPC-A';
+            //See if the serial number starts with 3 which means it is a drug and the middle 20 digits are the NDC.
```
Comment (ramseychambers) [inline lib/bark.js]: Lines 189, 190, 191, 192 shouldn't have been ported over.

File: bower.json (hunk 8 above)
Comment (nettej17) [reply bower.json]: yes, this was a deliberate mike change so that we use the component more correctly
